Sofa/framework/Geometry/src/sofa/geometry/Edge.h

```diff
@@ -121,6 +121,51 @@ struct Edge
     }
 
 
+    /**
+    * @brief	Test if a point is on Edge (n0, n1)
+    * @tparam   Node iterable container
+    * @tparam   T scalar
+    * @param	p0: position of the point to test
+    * @param	n0,n1: nodes of the edge
+    * @return	bool result if point is on Edge.
+    */
+    template<typename Node,
+        typename T = std::decay_t<decltype(*std::begin(std::declval<Node>()))>,
+        typename = std::enable_if_t<std::is_scalar_v<T>>
+    >
+        static constexpr bool isPointOnEdge(const Node& p0, const Node& n0, const Node& n1)
+    {
+        // 1. check if the point, n0 and n1 are aligned
+        const auto AB = n0 - n1;
+        const auto AC = n0 - p0;
+
+        T N2 = T(0);
+        if constexpr (std::is_same_v < Node, sofa::type::Vec<2, T> >)
+        {            
+            N2 = sofa::type::cross(AB, AC);
+        }
+        else
```

Comment:
This `else` will be triggered if Node is not Vec<2,T> so it is supposed to work with anything else. Which is what you dont want I guess.
you should test if Node is Vec<3,T> and if not, throw a static_assert() saying that this function is only usable with sofa::type<Vec<{2,3}, T>
Or directly do a static_assert at the beginning testing if Node is sofa::type<Vec<{2,3}, T> directly

Reply:
In fact if dot product is defined for Vec > 3d  the  method should be correct....
Not sure what is the good spirit to handle that. But for sure I need to avoid the Vec1 case.

Reply:
C++20 constraints!!!

Reply:
I let you do the suggestion 😄